### PR TITLE
Update docs/tutorials

### PR DESCRIPTION
**Pull request by ziyiyin97**
Open a PR so that @rafaelorozco and I could check the tutorials and make sure they are up to date

### CODE REVIEW

**Comment by mloubout** [comment]
Is this work in progress or did you finish the update?

**Comment by ziyiyin97** [comment]
Still in progress. Will finish it by the end of the week

File: src/TimeModeling/Utils/auxiliaryFunctions.jl [hunk 1]
```diff
@@ -444,7 +444,7 @@ function setup_3D_grid(xrec::Array{Array{T, 1}, 1},yrec::Array{Array{T, 1},1},zr
             for j=1:nxrec
                 xloc[i][idx] = xrec[i][j]
                 yloc[i][idx] = yrec[i][k]
-                zloc[i][idx] = zrec[i]
+                zloc[i][idx] = ( length(JUDI.tof32(zrec)) == 1 ? zrec[1] : zrec[i])
```
Comment (mloubout) [inline src/TimeModeling/Utils/auxiliaryFunctions.jl:447]: No the case "==1" should not be an array and should be caught line 466 for a single zrec value. If there is an `if` the multiple dispatch isn't set correctly or some weird input is there.

**Comment by mloubout** [comment]
status?

File: examples/scripts/modeling_basic_3D.jl [hunk 2]
```diff
@@ -7,7 +7,7 @@
 using JUDI
 
 # Set up model structure
-n = (120, 100, 90)    # (x,y,z) or (x,z)
+n = (20, 20, 20)    # (x,y,z) or (x,z)
```
Comment (mloubout) [inline examples/scripts/modeling_basic_3D.jl:10]: ?

File: src/TimeModeling/Utils/auxiliaryFunctions.jl [hunk 3]
```diff
@@ -728,6 +728,47 @@ function transducer(q::judiVector, d::Tuple, r::Number, theta)
...
+"""
+    SrcRecReciprocal(d_obs, q)
+
+The data d_obs is assumed to be generated by sequential fired sources q,
+SrcRecReciprocal outputs the data and source as if we switch the source/receiver locations thanks to reciprocity
+
+"""
+function SrcRecReciprocal(d_obs::judiVector, q::judiVector)
+
+    nsrc = length(d_obs.geometry.xloc[1])   # actual number of receivers
```
Comment (mloubout) [inline src/TimeModeling/Utils/auxiliaryFunctions.jl:741]: Need to check all sources see the same receivers.

File: src/TimeModeling/Utils/auxiliaryFunctions.jl [hunk 4]
```diff
@@ -728,6 +728,47 @@ function transducer(q::judiVector, d::Tuple, r::Number, theta)
...
+    nrec = deepcopy(d_obs.nsrc) # actual number of sources
```
Comment (mloubout) [inline src/TimeModeling/Utils/auxiliaryFunctions.jl:742]: It's a scalar, it's always a deep copy, not needed

File: src/TimeModeling/Utils/auxiliaryFunctions.jl [hunk 5]
```diff
@@ -728,6 +728,47 @@ function transducer(q::judiVector, d::Tuple, r::Number, theta)
...
+    
+    xsrc = convertToCell(d_obs.geometry.xloc[1])
+    ysrc = convertToCell(d_obs.geometry.yloc[1])
+    zsrc = convertToCell(d_obs.geometry.zloc[1])
+
+    src_geometry = Geometry(xsrc,ysrc,zsrc;dt=d_obs.geometry.dt[1],t=d_obs.geometry.t[1])
+
+    xrec = q.geometry.xloc[1]
+    yrec = q.geometry.yloc[1]
+    zrec = q.geometry.zloc[1]
```
Comment (mloubout) [inline src/TimeModeling/Utils/auxiliaryFunctions.jl:760]: That's only the first source position no the array of all source positions.

File: src/TimeModeling/Utils/auxiliaryFunctions.jl [hunk 6]
```diff
@@ -728,6 +728,47 @@ function transducer(q::judiVector, d::Tuple, r::Number, theta)
...
+
+    rec_geometry = Geometry(xrec,yrec,zrec;dt=q.geometry.dt[1],t=q.geometry.t[1])
```
Comment (mloubout) [inline src/TimeModeling/Utils/auxiliaryFunctions.jl:762]: need the `nsrc=nsrc` argument or assumes is single source

File: src/TimeModeling/Utils/auxiliaryFunctions.jl [hunk 7]
```diff
@@ -728,6 +728,47 @@ function transducer(q::judiVector, d::Tuple, r::Number, theta)
...
+
+    d_reci = judiVector(rec_geometry, reci_data)
+    q_reci = judiVector(src_geometry, q.wavelet)
```
Comment (mloubout) [inline src/TimeModeling/Utils/auxiliaryFunctions.jl:765]: `q.data`

File: src/TimeModeling/Utils/auxiliaryFunctions.jl [hunk 8]
```diff
@@ -728,6 +728,47 @@ function transducer(q::judiVector, d::Tuple, r::Number, theta)
     return judiVector(Geometry(xloc, yloc, zloc; t=t, dt=dt), data)
 end
 
+
+"""
+    SrcRecReciprocal(d_obs, q)
+
+The data d_obs is assumed to be generated by sequential fired sources q,
+SrcRecReciprocal outputs the data and source as if we switch the source/receiver locations thanks to reciprocity
+
```
Comment (mloubout) [inline src/TimeModeling/Utils/auxiliaryFunctions.jl:737]: Add argument list and types.

File: test/test_adjoint.jl [hunk 9]
```diff
@@ -10,6 +10,7 @@ parsed_args = parse_commandline()
 nlayer = parsed_args["nlayer"]
 tti = parsed_args["tti"]
 fs =  parsed_args["fs"]
+isic =  parsed_args["isic"]
```
Comment (mloubout) [inline test/test_adjoint.jl:13]: isic isn't one of the  command line arguments in any of the test setup so none of `isic=true` this is tested on;ly the default case is tested.

File: test/test_gradient_fwi.jl [hunk 10]
```diff
@@ -61,4 +61,23 @@ dt = srcGeometry.dt[1]
 	@test isapprox(rate_1, 1.25f0; rtol=5f-2)
 	@test isapprox(rate_2, 1.5625f0; rtol=5f-2)
 
-end
\ No newline at end of file
+end
+
+
+@testset "fwi_objective correct algebra test with $(nlayer) layers and tti $(tti) and freesurface $(fs) and isic $(isic)" begin
```
Comment (mloubout) [inline test/test_gradient_fwi.jl:67]: isic not tested as not part of the test suite setup, and the dft and checkpointing option are missing.

File: test/test_gradient_lsrtm.jl [hunk 11]
```diff
@@ -84,3 +85,47 @@ dt = srcGeometry.dt[1]
 	@test isapprox(Jls, Jfwi;rtol=0, atol=0)
 	@test isapprox(gradls, gradfwi;rtol=0, atol=0)
 end
+
+# Test if lsrtm_objective produces the same value/gradient as is done by the correct algebra
+
+for fs in [true, false]
```
Comment (mloubout) [inline test/test_gradient_lsrtm.jl:91]: Need to be a @testset

**Review by mloubout — COMMENTED** [review]
Thanks for the updates, need some fixing and proper setup if you want to setup isic like that.
Also please rename it to something more consistent with the comment since the doc is barely touched.

For future considertions (nitpicking)
-  it's not great practice to stack a bunch of different stuff in one PR. Make on PR per fix, it will also be a lot easier to merge and update/rebase/...
-  Try to name the commits according to its content so that it's easy to browse by commit. "local debug" aqnd such should not be in there and be skip in rebase

**Comment by mloubout** [comment]
ping

**Comment by ziyiyin97** [comment]
@mloubout could you provide with me some suggestions on how to make the test set? I am thinking of adding `isic` to the argument and to the `runtest.jl` file but what about dft/optimal checkpointing etc?

For example, `lsrtm_objective` w/ `isic=true` should pass the gradient test (Taylor series one). However, `lsrtm_objective` w/ 'optimal_checkpointing=true` should NOT pass the gradient test but only follow the correct algebra, i.e. objective function `= 0.5 * ||J^T(Jx-d)||_2^2`, gradient `=J^T(Jx-d)`. It would be kind of messy to put all of these to the argument list

**Comment by mloubout** [comment]
Why would it fail with checkpointing??? It's supposed to give the exact same gradient. Do you know what checkpointing does?

**Comment by ziyiyin97** [comment]
My bad. Checkpointing computes the true gradient. How about gradients computed by several frequencies?

**Comment by ziyiyin97** [comment]
Most of the things here were solved by previous PRs, e.g. #87 #64 #91 . I may open another one to address some remaining issues (if any). Closing.

**Comment by mloubout** [comment]
@rafaelorozco @ziyiyin97 one of the main goal of this PR was to update and fix the tutorials, examples and documentation. None of these happened in the end and just a few PRs adding additional un-documented features isn't really helping.

**Comment by ziyiyin97** [comment]
The leading order problem of the tutorial is about sim src, which is mentioned in #92 . It needs a bit work to implement a weighting operator. I will get back to it after SEG

**Pull request closed without merging**